File: src/lib/api/entities.ts
import { invoke } from "@tauri-apps/api/core";
import type { Entity, EntityPatch } from "./types";

export function createEntity(
  spaceId: string,
  type: string,
  title: string,
  icon: string | null,
): Promise<Entity> {
  return invoke("create_entity", { spaceId, type, title, icon });
}

export function getEntity(id: string): Promise<Entity> {
  return invoke("get_entity", { id });
}

export function updateEntity(id: string, patch: EntityPatch): Promise<Entity> {
  return invoke("update_entity", { id, patch });
}

export function listEntities(spaceId: string | null, includeDeleted: boolean): Promise<Entity[]> {
  return invoke("list_entities", { spaceId, includeDeleted });
}

export function softDeleteEntity(id: string): Promise<void> {
  return invoke("soft_delete_entity", { id });
}

export function restoreEntity(id: string): Promise<void> {
  return invoke("restore_entity", { id });
}

/// Permanently removes everything in the Trash; resolves to how many items went.
export function emptyTrash(): Promise<number> {
  return invoke("empty_trash");
}

export function hardDeleteEntity(id: string): Promise<void> {
  return invoke("hard_delete_entity", { id });
}

/// A copy in the same Space, titled "<title> (copy)", with the same fields, icon,
/// labels and block content.
export function duplicateEntity(id: string): Promise<Entity> {
  return invoke("duplicate_entity", { id });
}

/// Turns an entity into another type in place (same id, relationships and
/// labels), through the conversions the backend registers.
export function convertEntity(id: string, to: string): Promise<Entity> {
  return invoke("convert_entity", { id, to });
}

File: src-tauri/src/commands/entities.rs
use crate::db::entities::{self, Entity, EntityPatch};
use crate::db::DbState;
use crate::error::AppResult;
use tauri::State;

#[tauri::command]
pub fn create_entity(
    state: State<DbState>,
    space_id: String,
    r#type: String,
    title: String,
    icon: Option<String>,
) -> AppResult<Entity> {
    let conn = state.0.lock().unwrap();
    entities::create_entity(&conn, space_id, r#type, title, icon)
}

#[tauri::command]
pub fn list_entities(
    state: State<DbState>,
    space_id: Option<String>,
    include_deleted: bool,
) -> AppResult<Vec<Entity>> {
    let conn = state.0.lock().unwrap();
    entities::list_entities(&conn, space_id.as_deref(), include_deleted)
}

#[tauri::command]
pub fn update_entity(state: State<DbState>, id: String, patch: EntityPatch) -> AppResult<Entity> {
    let conn = state.0.lock().unwrap();
    entities::update_entity(&conn, &id, patch)
}

#[tauri::command]
pub fn get_entity(state: State<DbState>, id: String) -> AppResult<Entity> {
    let conn = state.0.lock().unwrap();
    entities::get_entity(&conn, &id)
}

#[tauri::command]
pub fn soft_delete_entity(state: State<DbState>, id: String) -> AppResult<()> {
    let conn = state.0.lock().unwrap();
    entities::soft_delete_entity(&conn, &id)
}

#[tauri::command]
pub fn restore_entity(state: State<DbState>, id: String) -> AppResult<()> {
    let conn = state.0.lock().unwrap();
    entities::restore_entity(&conn, &id)
}

#[tauri::command]
pub fn hard_delete_entity(state: State<DbState>, id: String) -> AppResult<()> {
    let conn = state.0.lock().unwrap();
    entities::hard_delete_entity(&conn, &id)
}

/// A copy of the entity in the same Space, through its registered schema.
#[tauri::command]
pub fn duplicate_entity(state: State<DbState>, id: String) -> AppResult<Entity> {
    let conn = state.0.lock().unwrap();
    let data = crate::db::schema::duplicate(&conn, &id)?;
    let new_id = crate::db::schema::payload_id(&data).ok_or_else(|| {
        crate::error::AppError::Db("internal: could not locate id in duplicate result".into())
    })?;
    entities::get_entity(&conn, &new_id)
}

/// Turns an entity into another type in place, through the conversions
/// registered in `schema` (the same ones the CLI's `convert` verb offers).
#[tauri::command]
pub fn convert_entity(
    state: State<DbState>,
    id: String,
    to: String,
) -> AppResult<crate::db::entities::Entity> {
    let conn = state.0.lock().unwrap();
    crate::db::schema::convert(&conn, &id, &to)?;
    crate::db::entities::get_entity(&conn, &id)
}
